### ISSUE TITLE
Example implementation of CRUD with REST JSON API

### ISSUE THREAD

**Issue by barrystaes** [acdlite/flummox#141]
I am asking a lot, i know. But i cant do this without you.

Its hard to decide on what Flux flavor to use for my SPA. Right now Flummox is my favorite due to being Isomorphic, relatively popular, in active development. However it comes short on example implementations.

Just about any Flux flavor has a TODO example, and most lack server communication. I can't even imagine someone building apps without that.. i'm flabbergasted as to why there are no examples on that. I understand there are many ways to have data move from API to stores, but i just need one. *I need it to work first, so i too can start looking at alternatives.*

So **i humbly ask to provide an example** that;
- has UserStore and actions by Flummox
- talks with REST API in JSON such as `/users/:id` with `GET`, `POST`, `PUT`, and `DELETE`.

I believe this wil help Flummox adaption, as this is what just about anyone wants to use it for.

**Comment by acdlite** [comment]
I feel ya. I would love to have more examples. Right now we only have the one, which is the docs app, but that only covers `GET`. There are a few other examples in the wild, but none of them are full CRUD.

An example with a UserStore (which I assume is like a SessionStore) is a great idea. I get a lot of questions about this.

I have something in mind. Can't make any promises on when it will be done, though.

**Comment by barrystaes** [comment]
I was **not** referring to a 'sessionstore' here, but a list of items to
create/read/update/delete. With the item type being something that many
SPAs need: User.

With some fields like:

- username,
- email,
- language (dropdown string),
- active (boolean)
- (readonly) number of logins (integer),
- (readonly) last login (iso8601 utc date) and a
- (write-only) password (perhaps with SHA in a future example iteration)

This demonstrates the various types of fields and different but realistic
data manipulation features.
Op 7 apr. 2015 18:13 schreef "Andrew Clark" <notifications@github.com>:

**Comment by barrystaes** [comment]
The example could be with fewer fields of course, but a realistic example has at least a string, number, and boolean field.

@acdlite what did you have in mind?

**Comment by acdlite** [comment]
Usually (often) CRUD apps have some soft of authentication. I've gotten many inquiries about how to do isomorphic authentication with Flux.

**Comment by barrystaes** [comment]
Yes that is another problem i too need to solve. Id be happy to such
example implementation where many eyes agree upon.
But its wont even matter if i cant CRUD in the first place. :) Auth seems a
way simpler problem once there is an app going.
Op 10 apr. 2015 14:43 schreef "Andrew Clark" <notifications@github.com>:

**Comment by acdlite** [comment]
Here's an isomorphic example with Rails by @nambrot:

https://github.com/nambrot/rails-webpack-react-flux

I'll make a list of these as I become aware of them

**Comment by barrystaes** [comment]
Sorry i do not see how that example is relevant. Its not using a REST API and its not a full CRUD either. 

I'm asking insights in how to have stores to keep track of multiple fields per item (the right way), and in how to have stores (or anything really) talk to a serverside REST API.

There obviously are a lot of best practices on that, but i dont know them yet - but i would like to.

**Comment by acdlite** [comment]
You're right, it's not exactly what you asked for, but I'm posting it here for anyone else who's looking for examples that might be helpful.

If I had unlimited time I would have already created the example you're asking for, plus several others. But I don't, so please be patient. I'll have something for you as soon as I get to it.

**Comment by nambrot** [comment]
Sorry to just break in, but yes my example does use a serverside REST API. Admittedly, I've not touched associations and user authentication.

**Comment by barrystaes** [comment]
I stand corrected, sorry @nambrot. :+1: 
I might have misread your list of cons on other examples as if they where features of yours.

**Comment by nambrot** [comment]
No Problem, as said I feel your frustration. Let me know if you have any questions.

**Comment by jirivrany** [comment]
Hi Barry. When I started learning flummox, I found an example made by @vesparny - https://github.com/vesparny/flux-immutable-example It is not complete CRUD, it only Reads the data from API mock. But maybe you will find it handy.

**Comment by barrystaes** [comment]
Thanks @jirivrany i will look into that monday.

I was going about making an CRUD API example myself, and then ask some gurus to tell me what i did wrong.. ;) https://github.com/barrystaes/react-starter/tree/userstore-flummox?files=1

**Comment by barrystaes** [comment]
I gave up on making this CRUD example, and also on using Flummox.

If anyone feels like showing how its done with Flummox, i'd be happy to hear!

**Comment by vesparny** [comment]
@barrystaes take a look at my github.
I handled basic http calls wit express/flummox here:
https://github.com/vesparny/widget

**Comment by alduro** [comment]
@barrystaes great sample. I'm using it. Thanks.

**Comment by swordsreversed** [comment]
I've looked at @vesparny's repo and I can't see any C, U or D of CRUD in here. Could someone point it out?

**Comment by eranimo** [comment]
I'm trying to build an isomorphic ES7 application using Flummox and I'm having trouble with auth. I used the docs as a base, using koa as a server to render react. I am also using a proxy to proxy api calls to my  api server located on a different port. The problem is, I can't get my proxy server to correctly pass cookies to the client so that it can authenticate.

I ended up putting a JWT in the Authentication header, but that introduces new problems. I have a wrapper component that checks to see if you're logged in. If you're not, the auth store redirects you to the login page. If you have a token in your cookie, it tries to authenticate it by making an api call with the token in the Authentication header. The problem is, this is happening in the static function (just like in the docs site) which enables api calls to work isomorphically, but obviously the server doesn't have cookies, so I have to pass in the headers to that static function. This basically means I have to pass the auth token as a parameter to every api call.

An example doing auth with an API, through a proxy would be fantastic.

**Comment by alduro** [comment]
Same as @eranimo here. Using React-router I can not make auth working checking whether user is logged and redirect to login if not. More then 3 days going around on this.
